### DISCUSSION TITLE
Accept ours / theirs works on conflict stage per file

### DISCUSSION THREAD

**Discussion by GlebHmarsky** [gitextensions/gitextensions#11788] [Ideas] [answered]
Hi,

### Why

Let me first explain the need for this feature. My team and I use Git Extensions for Unreal Engine 4 projects. All files that Unreal Engine uses are binaries. The content looks like this:

```
version https://git-lfs.github.com/spec/v1
oid sha256:2f2d30a9be890c01019009a31d8830e0b0a15106c1964889def90b2fbb9e3953
size 1517638
```

It's just a hash with some size in bytes. 
Since all our assets look like this, it is effectively impossible to do any meaningful conflict resolution.
Now, imagine we have a conflict over a file named `ActorClass.uasset`. This puts us in a situation with only two options:

1. Reset our work in `ActorClass.uasset` and accept someone else's changes, then redo the changes you've lost manually in Unreal Engine once again from scratch.
2. Override someone else's changes in `ActorClass.uasset` and prioritize ours.

### The Idea

My team and I are missing the ability to right-click on a conflict file and select the desired action.

I've researched Git commands that can help with this:

1. `git checkout --theirs path/to/file` - Accept **incoming** changes for the file.
2. `git checkout --ours path/to/file` - Accept **local** changes for the file.

### Suggestion

When encountering a conflict, you go to the "Commit" window and see files with merge conflicts. By right-clicking (RMB) on a file, there would be two buttons in the context menu: "Accept incoming changes" and "Accept local changes."

I’ve provided a picture of the exact context menu window I mean:

![image](https://github.com/gitextensions/gitextensions/assets/62330322/e73de6a1-d5cd-4956-93a0-2d68d28e3c47)

Each button would execute the corresponding commands mentioned in the "Idea" section.

**Comment by RussKie** [comment]
I spent about 6 months a project where I worked a on cross-cutting changes for a very busy repo. Naturally rebasing was very painful, and I had to deal with a lot of pain your trying to deal now. So, I'm fully supportive of the ask 

To that extent I even started work that would enable custom merge strategy selection in dialogs (pull/rebase/etc.), but (as usual) I ran out of time I could afford to this...

**Comment by gerhardol** [comment]
Not disagreeing to have conflict resolution in the commit dialog (as well as Diff tab), but use Solve merge conflicts (there is a button in the grid when there are conflicts).
(It might be that resolve for multiple files are not handled well.)

![image](https://github.com/gitextensions/gitextensions/assets/6248932/5634c772-e549-4e57-8b96-17b54c633438)

**Reply by RussKie** [reply]
Yep, this is a function of that dialog IMO

**Reply by GlebHmarsky** [reply]
It's a good point. Let me explain the problem in more detail. You see, all the merge tools that Git Extensions knows about (in the settings merge tool field) do not fit my expectations:

Here is a picture of them:
![image](https://github.com/gitextensions/gitextensions/assets/62330322/1826fb6b-9fbd-4d31-af74-0d206df3bb4c)

Additionally, I have tried many others, like:
- fs
- bindiff
- vbindiff

These tools can open files and even show me the differences between them, but if I decide to select all changes from one file, it would corrupt my files. As in the example above, my `ActorClass.uasset` would not be visible in the project.

Here's a forum post in the Unreal Engine section about merging problems, which mentions only two ways to handle it, exactly as I described:
[Unreal Engine Forum - Blueprint Merge Tool](https://forums.unrealengine.com/t/4-27-any-blueprint-merge-tool-out-there/502874)

So, I would really love to have those two commands implemented, maybe even in the "Solve Merge Conflict" window for better per-file readability.

**Reply by gerhardol** [reply]
> So, I would really love to have those two commands implemented, maybe even in the "Solve Merge Conflict" window for better per-file readability.

It is there already (but maybe needs tuning for many files?).
https://git-extensions-documentation.readthedocs.io/en/main/modify_history.html#handle-merge-conflicts

**Reply by GlebHmarsky** [reply, marked as answer]
Oh, i feel so relief and stupid right now :)
This is exactly that i was looking for! 
I will make this comment as Answer and explain for whoever follows, but all thanks to @gerhardol 

I didn't know that you can click with RMB on conflict file and there will appear context menu with  necessary buttons to Choose ours/theirs/base
Here is a picture
![image](https://github.com/gitextensions/gitextensions/assets/62330322/973ce248-7dd6-4091-8ccb-8466edec8eb1)

P.S.
> but maybe needs tuning for many files?

I think that is unnecessary cause you have option to select multiple files at once and choose same 3 options 
Maybe for pull options would be nice to have `--ours`  `--theirs` keys access, but i think this is another conversation, because i don't have good cases where i could use them more often. With "Solve Merge Conflict" and RMB me and my team now able to cover our needs
